test/exe/admin_response_test.cc

```diff
@@ -102,10 +102,24 @@ class AdminStreamingTest : public AdminRequestTestBase, public testing::Test {
    *
    * @param url the admin endpoint to initiate.
    */
-  void blockMainThreadUntilResume(absl::string_view url, absl::string_view method) {
+  /*void blockMainThreadUntilResume(absl::string_view url, absl::string_view method) {
     AdminResponseSharedPtr blocked_response = main_common_->adminRequest(url, method);
     blocked_response->getHeaders(
         [this](Http::Code, Http::ResponseHeaderMap&) { resume_.WaitForNotification(); });
+        }*/
+
+  void interlockMainThread(std::function<void()> fn,
+                           std::function<void()> run_before_resume = nullptr) {
+    main_common_->dispatcherForTest().post([this, fn] {
+      resume_.WaitForNotification();
```

Comment:
I'd suggest still adding another sync here to make sure the main thread has *entered* the posted block by the time you call run_before_resume(). Otherwise there might still be two dispatcher posts queued up before this one that run_before_resume could execute prior to, which would most likely be surprising.
i.e.
```
    absl::Notification main_thread_reached_interlock;
    main_common_->dispatcherForTest().post([this, fn, &main_thread_reached_interlock] {
      main_thread_reached_interlock.Notify();
      resume_.WaitForNotification();
      fn();
      pause_point_.Notify();
    });
    main_thread_reached_interlock.WaitForNotification();
    [...]
```

Reply:
Ah OK I wound up putting in this run_before_resume hack and then, as it was only used for the timeout case, removing it.

I think given the single use of run_before_resume you saw I don't need yet another notification. However I'm going to look at using ThreadSynchronizer now as it's a little cleaner than using 2 notifications.

Reply:
Yep, without `run_before_resume` it's fine as the dispatcher itself effectively acts as the "resolve all previous things" synchronization.